src/utils/algorithm_specifiers.rs

```diff
@@ -0,0 +1,699 @@
+    }
+}
+
+impl TryFrom<TPM2_ALG_ID> for ObjectType {
+    type Error = Error;
+
+    fn try_from(algorithm_id: TPM2_ALG_ID) -> Result<Self> {
+        match algorithm_id {
+            TPM2_ALG_NULL => Ok(ObjectType::Null),
+            TPM2_ALG_RSA => Ok(ObjectType::Rsa),
+            TPM2_ALG_ECC => Ok(ObjectType::Ecc),
+            TPM2_ALG_KEYEDHASH => Ok(ObjectType::KeyedHash),
+            TPM2_ALG_SYMCIPHER => Ok(ObjectType::SymCipher),
+            _ => Err(Error::local_error(WrapperErrorKind::InvalidParam)),
+        }
+    }
+}
+
+////////////////////////////////////////////////
+///
+/// Asymmetric algorithms.
+///
+/// The specification specifies these as
+/// "asymmetric algorithm with a public and private key".
+////////////////////////////////////////////////
+
+/// Enum containing asymmetric algorithms.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum AsymmetricAlgorithm {
+    Rsa,
+    Ecc,
+}
+
+impl From<AsymmetricAlgorithm> for TPM2_ALG_ID {
+    fn from(asymmetric_algorithm: AsymmetricAlgorithm) -> Self {
+        match asymmetric_algorithm {
+            AsymmetricAlgorithm::Rsa => TPM2_ALG_RSA,
+            AsymmetricAlgorithm::Ecc => TPM2_ALG_ECC,
+        }
+    }
+}
+
+impl TryFrom<TPM2_ALG_ID> for AsymmetricAlgorithm {
+    type Error = Error;
+
+    fn try_from(algorithm_id: TPM2_ALG_ID) -> Result<Self> {
+        match algorithm_id {
+            TPM2_ALG_RSA => Ok(AsymmetricAlgorithm::Rsa),
+            TPM2_ALG_ECC => Ok(AsymmetricAlgorithm::Ecc),
+            _ => Err(Error::local_error(WrapperErrorKind::InvalidParam)),
+        }
+    }
+}
+
+////////////////////////////////////////////////
+///
+/// Keyed hash (KEYEDHASH)
+///
+////////////////////////////////////////////////
+
+/// Enum containing keyed hash
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum KeyedHash {
+    Hmac,
+    Xor,
+}
+
+impl KeyedHash {
+    /// Returns the TPM2_ALG_ID that corresponds to the
+    /// keyed hash.
+    pub fn alg_id(self) -> TPM2_ALG_ID {
+        match self {
+            KeyedHash::Hmac => TPM2_ALG_HMAC,
+            KeyedHash::Xor => TPM2_ALG_XOR,
+        }
+    }
+}
+
+impl From<KeyedHash> for TPM2_ALG_ID {
+    fn from(keyed_hash: KeyedHash) -> Self {
+        match keyed_hash {
+            KeyedHash::Hmac => TPM2_ALG_HMAC,
+            KeyedHash::Xor => TPM2_ALG_XOR,
+        }
+    }
+}
+
+impl TryFrom<TPM2_ALG_ID> for KeyedHash {
+    type Error = Error;
+
+    fn try_from(algorithm_id: TPM2_ALG_ID) -> Result<Self> {
+        match algorithm_id {
+            TPM2_ALG_HMAC => Ok(KeyedHash::Hmac),
+            TPM2_ALG_XOR => Ok(KeyedHash::Xor),
+            _ => Err(Error::local_error(WrapperErrorKind::InvalidParam)),
+        }
+    }
+}
+
+////////////////////////////////////////////////
+///
+/// Symmetric algorithms.(SYMCIPHER)
+///
+////////////////////////////////////////////////
+
+/// Enum containing symmetric algorithms.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum SymmetricAlgorithm {
```

Comment:
I'm not sure this is needed - i.e. I'm not entirely sure where you would need this instead of the `Cipher` enum. Do you have an intended use-case where having just the variant has a compelling advantage?

My (rather limited) experience is that you'd want to have some way to identify the "whole" encryption algorithm so that you can represent it in a `TPMT_SYM_DEF` or `TPMT_SYM_DEF_OBJECT` structure.

That one doesn't allow conversion from `TPM2_ALG_ID`, but I'm not sure that along would be useful.

Reply:
You are probably right. I just looked at the table in the specification and thought I would make something that l resembled that(names and stuff.). 

SymmetricAlgorithms is intended to be connected with the ObjectType SymCipher. If I am not misstaken it should contain the only values that is possible for that object type. But that is parhaps a little backward thinking. I think Cipher is doing it in the right way. You select a Cipher and get a correct ObjectType.

Reply:
So the only scenario I could think of would be for some parsing. Where you would like to check the correctness of a TPMT_PUBLIC. But that is parhaps an extremly far fetched reason to leave t?

Reply:
>So the only scenario I could think of would be for some parsing. Where you would like to check the correctness of a TPMT_PUBLIC. 

Ok, I think that makes sense - even more so for `TPMT_SENSITIVE` where the algorithm information isn't duplicated. 

I'm ok with this going in as is for now, I'll fix the `Cipher` later to make it more sensible.